### PR TITLE
fix: add module to OpenAPI schema key names

### LINKED ISSUE #2471: enhancement: OpenAPI re-use existing schemas of same name
### Description

We're getting an unexpected `ImproperlyConfiguredException` in `litestar/_openapi/schema_generation/schema.py", line 594, in process_schema_result`.

In our app, we defined a base`Player` model using Pydantic. This model is returned on its own and also as a subdocument in other endpoints. 

Schema generation fails recognizes two different schemas with the same title. However, the schemas only differ in their description. The description itself is set as a field-level annotation when `Player` is used as a subdocument. 

This makes setting the `description` at the field level much less useful.

I may be missing something. Looking at the code it hashes the schemas to ensure a 1-1 relationship between titles and schemas. I don't believe this was previously a constraint and schemas were produced based on the module path. I may not understand why this is now a requirement.

I've provided an example showing this behavior for both msgspec and Pydantic.

This behavior is new to Litestar 2.0 and did not exist in Litestar 1.51

### URL to code causing the issue

_No response_

### MCVE

```python
### Example showing the issue with Pydantic

import traceback
from litestar import Litestar, get
from litestar.params import Parameter
from pydantic import BaseModel, Field

from litestar.types.asgi_types import Scope


class Player(BaseModel):
    id: str
    name: str
    position: str
    jersey: str


class Play(BaseModel):
    id: str
    ball_carrier: Player = Field(description="Sub document representing player who carried the ball")


@get("/player", sync_to_thread=True)
def get_player_by_id(id: str = Parameter(str)) -> Player:
    return Player(id=id, name="Chris", position="RB", jersey="00")


@get("/play", sync_to_thread=True, include_in_schema=True)
def get_play_by_id(id: str = Parameter(str)) -> Play:
    return Play(id=id, ball_carrier=Player(id="asdasd", name="chris", position="rb", jersey="00"))


async def after_exception_hook(exc: Exception, scope: "Scope") -> None:
    traceback.print_exc()


app = Litestar(route_handlers=[get_player_by_id, get_play_by_id], after_exception=[after_exception_hook])

if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app)
```

### Example using msgspec

```py
import traceback
from litestar import Litestar, get
from litestar.params import Parameter

from litestar.types.asgi_types import Scope

from msgspec import Meta, Struct
from typing import Annotated


class Player(Struct):
    id: str
    name: str
    position: str
    jersey: str


class Play(Struct):
    id: str
    ball_carrier: Annotated[Player, Meta(description="Sub document representing player who carried the ball")]


@get("/player", sync_to_thread=True)
def get_player_by_id(id: str = Parameter(str)) -> Player:
    return Player(id=id, name="Chris", position="RB", jersey="00")


@get("/play", sync_to_thread=True, include_in_schema=True)
def get_play_by_id(id: str = Parameter(str)) -> Play:
    return Play(id=id, ball_carrier=Player(id="asdasd", name="chris", position="rb", jersey="00"))


async def after_exception_hook(exc: Exception, scope: "Scope") -> None:
    traceback.print_exc()


app = Litestar(route_handlers=[get_player_by_id, get_play_by_id], after_exception=[after_exception_hook])

if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app)

```
```


### Steps to reproduce

```bash
1. Create a base app
2. Define a model and set it as the return type of an endpoint
3. Define a second model, containing a field annotated with the model defined in step 2. Set a `description` on the field (i.e. Pydantic `Field(description="{desc}"` or msgspec  Annotated[Player, Meta(description="{desc}")]). Create a second endpoint returning this model
4. See error
```

### Screenshots

_No response_

### Logs

```bash
litestar.exceptions.http_exceptions.ImproperlyConfiguredException: 500: Two different schemas with the title Player have been defined.

first: {"properties":{"id":{"type":"string"},"name":{"type":"string"},"position":{"type":"string"},"jersey":{"type":"string"}},"type":"object","required":["id","jersey","name","position"],"title":"Player"}
second: {"properties":{"id":{"type":"string"},"name":{"type":"string"},"position":{"type":"string"},"jersey":{"type":"string"}},"type":"object","required":["id","jersey","name","position"],"title":"Player","description":"Sub document representing player who carried the ball"}
```

### Litestar Version

Python 3.11
litestar[full] @ 2.2.1
pydantic @ 2.4.2

### Platform

- [ ] Linux
- [X] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh Litestar dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/2471">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2471/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by chris-telemetry**
### Pull Request Checklist

- [x] New code has 100% test coverage
- [x] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [x] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR
- [x] Pre-Commit Checks were ran and passed
- [x] Tests were ran and passed

### Description
Addressing issue raised [here](https://github.com/litestar-org/litestar/issues/2471). Change schema keys to include full path information instead of simply using `__name__`. This allows model reuse while maintaining OpenAPI schema generation functionality.

-

### Close Issue(s)
Closes #2471

-

### CODE REVIEW

File: litestar/_openapi/schema_generation/utils.py [hunk 1]
```diff
@@ -15,3 +17,16 @@ def sort_schemas_and_references(values: list[Schema | Reference]) -> list[Schema
...
+    """Normalize a type annotation, replacing all non-alphanumeric with underscores. Existing underscores will be left as-is
+
+    Args:
+        type_annotation_str (str): A string representing a type annotation (i.e. 'typing.Dict[str, typing.Any]' or '<class 'model.Foo'>')
+
+    Returns:
+        str: A normalized version of the input string
+    """
+    # Use a regular expression to replace non-alphanumeric characters with underscores
+    return re.sub(r"[^a-zA-Z0-9_]", "_", type_annotation_str)
```
[the remark this answers is not in the document]
Comment (peterschutt) [reply litestar/_openapi/schema_generation/utils.py:32]: Also, we should have an explicit unit test for this function.

**Review by peterschutt — COMMENTED** [review]
Hey, thanks for this! I've made some comments/suggestions in-line.

To summarize the changes, we are modifying the key part of the path to the schemas of the objects, but the schemas themselves remain unchanged. Is that a fair characterization?

Can you comment on whether you think this should be considered breaking current behavior, or not and why/why not? If breaking, it doesn't mean we cannot get these changes in 2.x, we might just have to put the behavior behind a feature flag.

File: tests/examples/test_openapi.py [hunk 2]
```diff
@@ -29,7 +31,7 @@ def test_schema_generation() -> None:
             },
             "components": {
                 "schemas": {
-                    "IdContainer": {
+                    container_key: {
```
[the remark this answers is not in the document]
Comment (chris-telemetry) [reply tests/examples/test_openapi.py:34]: That's correct!

File: tests/examples/test_openapi.py [hunk 3]
```diff
@@ -1,10 +1,12 @@
 from docs.examples.openapi import customize_pydantic_model_name
 
+from litestar._openapi.schema_generation.utils import normalize_type_name
 from litestar.testing import TestClient
 
 
 def test_schema_generation() -> None:
     with TestClient(app=customize_pydantic_model_name.app) as client:
+        container_key = normalize_type_name(str(customize_pydantic_model_name.IdModel))
```
[the remark this answers is not in the document]
Comment (chris-telemetry) [reply tests/examples/test_openapi.py]: Updated to use the literal wherever possible. Some places it's not -- such as where `pytest.mark.parameterize` is being used.

File: litestar/_openapi/schema_generation/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (chris-telemetry) [reply litestar/_openapi/schema_generation/utils.py]: Wrote one here:
`./tests/unit/test_openapi/test_schema.py::test_get_normalized_schema_key`

**Comment by chris-telemetry** [comment]
@peterschutt I've made updates per your comment. I also made some updates to reflect changes made in the interim to support generic OpenAPI schema definitions. 

Sorry this took awhile -- went on vacation and totally forgot about it!

**Comment by chris-telemetry** [comment]
>To summarize the changes, we are modifying the key part of the path to the schemas of the objects, but the schemas themselves remain unchanged. Is that a fair characterization?

That's correct!



> Can you comment on whether you think this should be considered breaking current behavior, or not and why/why not? If breaking, it doesn't mean we cannot get these changes in 2.x, we might just have to put the behavior behind a feature flag.

I wouldn't think it should be breaking for the vast majority of users. It's of course impossible to know how everybody is using this functionality, so I can't say that with 100% certainty. Perhaps some people are relying on how their keys are named currently -- hard to know for certain. I've tested it using Postman's collection import (before and after) and it seems to be just fine. That's the only use-case I can think of where the key name is relevant.

I know you've marked this as an enhancement and I do understand that line of thinking, but in my view this is more along the lines of a bug/oversight as it was supported in Starlite 1.0 and now no longer works. In fact, I found it while migrating one of our products from 1.0 to 2.0.

**Review by guacs — APPROVED** [review]
I think this looks good and I agree with you that this shouldn't be considered breaking. I'd like to get @peterschutt's opinion on this before merging this.

File: litestar/_openapi/schema_generation/utils.py [hunk 4]
```diff
@@ -82,6 +84,22 @@ def _should_create_literal_schema(field_definition: FieldDefinition) -> bool:
...
+    Args:
+        type_annotation_str (str): A string representing a type annotation (i.e. 'typing.Dict[str, typing.Any]' or '<class 'model.Foo'>')
+
+    Returns:
+        str: A normalized version of the input string
+    """
+    # Use a regular expression to replace non-alphanumeric characters with underscores
+    return re.sub(TYPE_NAME_NORMALIZATION_REGEX, "_", type_annotation_str)
+
+
```
Comment (peterschutt) [inline litestar/_openapi/schema_generation/utils.py:102]:
```suggestion


```

**Review by peterschutt — APPROVED** [review]
OK, thanks for your patience!

I'm happy to proceed with this as a bug-fix, however I think there is risk that we break stuff downstream. Accessing a schema by key, and testing against it is a common thing that we do in our own tests, so not unreasonable to think it would be happening downstream, and who knows what else. Let's see what arises and be attentive to any issues if they surface.

I had to fix one test after merging the latest changes in, which made me realize that using the normalized keys was a bit cumbersome. It was difficult to know where there should be double-underscores, leading and trailing underscores, and the word "class" felt redundant. So I slipped through one last commit that modifies the normalization function. Where previously the key might have been "_class__something_Model__" it will now be "something_Model". Please take one final look at that @chris-telemetry @guacs and LMK if you like/dislike/foresee issues.

Otherwise, I'm happy to proceed.

Thanks for your efforts on this @chris-telemetry!

**Comment by peterschutt** [comment]
> @peterschutt I noticed that you're working on this. I just wanted to point out that maybe this should be considered a feature instead of a fix. I feel that while it may have existed in starlite as @chris-telemetry pointed out, as far litestar 2 is considered, this is a new feature.
> 
> I don't want to speak for him, but I think @provinzkraut may share the same/similar thoughts.

Hmm, well as you can see in my review, I let myself be swayed but I can foresee there being issues.

**Review by guacs — APPROVED** [review]
@peterschutt I'm happy with the new changes regarding the naming you've introduced. The names are far more readable :)

**Comment by guacs** [comment]
> Accessing a schema by key, and testing against it is a common thing that we do in our own tests, so not unreasonable to think it would be happening downstream, and who knows what else. Let's see what arises and be attentive to any issues if they surface.

I hadn't considered that users could be doing that in their tests as well. I agree that we should be attentive to any possible issues and if it is causing too much trouble, we could consider reverting this PR at that point.

**Pull request merged**